docs/rfcs/20220804-ptensor/README.md

```diff
@@ -0,0 +1,215 @@
+* Linear Algebra
+  * `matmul(rhs, lhs) : (ptensor.ptensor, ptensor.ptensor) -> ptensor.ptensor`
+  * `matrix_transpose(rhs) : (ptensor.ptensor) -> ptensor.ptensor`
+  * `vecdot(rhs, lhs, axis) : (ptensor.ptensor, ptensor.ptensor, int64) -> ptensor.ptensor`
+  * `tensor_dot(rhs, lhs, axis) : (ptensor.ptensor, ptensor.ptensor, array) -> ptensor.ptensor`
+* Searching
+  * `find_index{$fop}(rhs, axis) : (ptensor.ptensor, int64, bool) -> ptensor.ptensor`
+    * `$fop = ['argmax', 'argmin']`
+  * `nonzero(rhs) : (ptensor.ptensor) -> tuple`
+  * `where(cond, rhs, lhs) : (scf.condition, ptensor.ptensor, ptensor.ptensor) -> ptensor.ptensor`
+* Set Functions
+  * `unique_all(rhs) : (ptensor.ptensor) -> [ptensor.ptensor, ptensor.ptensor, ptensor.ptensor, ptensor.ptensor]`
+  * `unique_counts(rhs) : (ptensor.ptensor) -> [ptensor.ptensor, ptensor.ptensor]`
+  * `unique_inverse(rhs) : (ptensor.ptensor) -> [ptensor.ptensor, ptensor.ptensor]`
+  * `unique_values(rhs) : (ptensor.ptensor) -> ptensor.ptensor`
+* Sorting Functions
+  * `sort{$sop}(rhs, descending, stable) : (ptensor.ptensor, bool, bool) -> ptensor.ptensor`
+  * `argsort{$sop}(rhs, descending, stable) : (ptensor.ptensor, bool, bool) -> ptensor.ptensor`
+* Statistical Functions
+  * `reduce{$rop}(rhs, axis, correction) : (ptensor.ptensor, int64) -> ptensor.ptensor`
+    * `$rop = ['max', 'min', 'mean', 'prod', 'sum', 'var', 'std']`
+* Utility Functions
+  * `test{$top}(rhs, axis) : (ptensor.ptensor, int64) -> ptensor.ptensor`
+    * `$rop = ['any', 'all']`
+
+### __Dist__ Operations
+The Dist dialect provides operations dealing with tensors which are partitioned and distributed across multiple processes. The operations assume soem kind of a runtime which handles aspects like communication and partitioning.
+- `init_dtensor(shape) : (ValueRange) -> (int64)`
+- `fini_dtensor(dtensor_id) : (int64) -> void`
+- `init_view(dtensor_id, slice) : (int64, slice) -> (int64)`
+- `local_shape(dtensor_id) : (int64) -> shape.shape`
+- `local_slice(dtensor_id) : (int64) -> slice`
+- `copy(from_id, to_id, from_ltensor, to_ltensor) : (int64, int64, RankedTensor, RankedTensor) -> void`
+- `finalize_reduce(dtensor_id, ltensor) : (int64, RankedTensor) -> void`
+
+### __CFD__ Operations
+The CFD dialect provides region operation which defines the target on which operations within the region should be executed on. The region is configureted with an attribute which defines the HW type, such as 'dist' for distributed, or 'CUDA', 'L0', and alike for GPU execution. The exact syntax and configuration possiblities depend on the targeted runtime.
+Regions can be nested.
+Additionally, regions accept an optional runtime parameter
+* a team identifier for distributed regions refering to a team of processes
+* a device identifier for GPU regions refering to a specific HW device
+
+Example nesting a GPU region into a distributed region ('IDR'= Intel Distributed Runtime)
+```
+cfd.region{'IDR'}(0) {
```

Comment:
I don't think 'CFD' is a good name for this, as is is not tied specifically to `compute follows data` concept, and op itself should better be named something like `device.region`

Reply:
What do you suggest instead of 'cfd'?

Reply:
At least in main I'm going to put it into `util` dialect as 'util' is not going away any time soon.

Reply:
'util' does not seem to be any better, so I assume you do not suggest this for refactor. In clean new world, what name would you prefer?

Reply:
Having the region op is a clean way to annotate the device on which a ptensor op will execute. 

But, I am not in favour of separate a `CFD` dialect. I feel the op should be inside the `Dist` dialect and `Intel GPU/Device/XPU` dialect. Something like:
```
dist.distributed_region
ixpu.device_region
```
The names are indicative and we can look for alternatives.

However, if these ops cannot be made part of `Dist` and  `Intel GPU/Device/XPU`, then the op should be part of `ptensor`.

Reply:
thanks @diptorupd @Hardcode84 I pushed a new revision without a cfd dialect and no dist_region.

Reply:
> dist.distributed_region
ixpu.device_region

I don't think having multiple versions of this op in different dialects is a good idea. We will need to copypaste a lot of canonicalization and support code between them.

Reply:
Also, I'm considering to use op for CPU side parallelism too (prange) but I didn't decided yet.

Reply:
Please note: the latest revision no longer has a distributed region. The region is mostly useful to carry the device information to lower passes and so enable cfd. We don't need this for dist.